### DISCUSSION TITLE
AVATTO S06 Pro Tuya

### DISCUSSION THREAD

**Discussion by interkom** [arendst/Tasmota#12741] [Q&A] [answered]
Good afternoon. I bought an infrared WiFi remote control AVATTO S06 Pro Tuya with temperature and humidity sensors.
https://aliexpress.ru/item/33018851576.html?spm=a2g39.orderlist.0.0.720d4aa6pCBhCx&_ga=2.110608197.1912497419.1627074659-308809112.1626529545&_gac=1.90974568.1626617825.CjwKCAjwos-HBhB3EiwAe4xM97TpWEJgVTpFrDazSRVloMIicDJEashcFIDK877XjSAAaOnEFl3L0BoCyL4QAvD_BwE&item_id=33018851576&sku_id=12000021757521239
I can not find information on it on the network. Is it supported by Tasmota?
Exactly the same without temperature and humidity sensors, I upgraded to Tasmota and ESPHOME, but I can't flash with temperature and humidity sensors.
I can't even figure out which microcircuit is connected to the ESP8266 and on which bus? What kind of temperature and humidity sensor is installed in this device?
Nothing is written on the microcircuit, but it is written on the sensor, but very small, I could only read AH10 A12113. If anyone knows where to get the documentation for this device or how to flash it so that the temperature and humidity are maintained, send a link.

**Comment by sfromis** [comment, marked as answer]
Devices known to be Tasmota supported are listed in the [templates repository](https://templates.blakadder.com/).

I do not recognize that sensor name, closest I know is the I2C temp/hum sensor AHT10, which is also what's stamped in the metal can of one I have. Thus it does not appear to be what you have there. Many sensors are I2C, but several relevant ones are not.

Tasmota supports a number of pretty standard IR pucks, but none with a T/H sensor. "Someone" may have to work out how your device is built, what pins are used, and if the sensor is a supported one (or maybe create a driver if not).

**Reply by interkom** [reply]
Thank you very much

**Comment by nikvostro** [comment]
Hope my experience below can help you.

I've bought [this one](https://www.aliexpress.com/item/1005002505900746.html?aff_fcid=1c89e58b30fc4c48be3399635d554b08-1627394404997-01625-_pvOzfGS&aff_fsk=_pvOzfGS&af=%7B%22c%22%3A%22aehYn2niwHMQyElx1610218081%22%7D&aff_platform=api-new-link-generate&sk=_pvOzfGS&aff_trace_key=1c89e58b30fc4c48be3399635d554b08-1627394404997-01625-_pvOzfGS&terminal_id=7f09ad3b7b1a4cfca1a54678806efa41) and obtained WB3S chip instead of ESP8266.
![image](https://user-images.githubusercontent.com/83143316/127168030-7d69e2e7-9f56-476f-8317-de26bf3234a9.png)

Here is the pins description I've researched:
![image](https://user-images.githubusercontent.com/83143316/127168434-c3ce8089-c78f-432d-8d66-55e7ccf8bf65.png)


Replaced WB3S it with ESP12-F:
![IR_hacked](https://user-images.githubusercontent.com/83143316/127168461-20b48f56-16a6-4e77-bc4a-8f84b9c94ab6.jpg)

There is a AHT10 temperature/humidity sensor is used (bottom left on pic). I've removed TuyaMCU chip and closed SCL/SDA pins on PCB directly to ESP8266. It works great with AHT1x Tasmota driver `#define USE_AHT1x` for my 2 Avatto devices:
![image](https://user-images.githubusercontent.com/83143316/127168757-dbda2c64-b50c-4166-bc40-f21f4ee92c7d.png)

![image](https://user-images.githubusercontent.com/83143316/127169568-17c8359b-0f3f-4c0d-93b1-d861eddd66c1.png)

**Reply by interkom** [reply]
Thank you very much. I will try it.

**Reply by interkom** [reply]
Thanks again. Soldered, flashed on ESPHOME and connected to Home Assistant

**Reply by steveboak** [reply]
How did you get IR and I2C both working at the same time? Currently using Tasmota 10.1.0, and I can get I2C working in standard tasmota.bin, or Fujitsu IR in tasmota_ir.bin, but not both together. Maybe you used tasmota_ircustom, which is now deleted?

**Reply by sfromis** [reply]
You can do your own [build](https://tasmota.github.io/docs/Compile-your-build/) including the feature mix you want. Beware that the size of a binary to include full IR along with other sensor drivers may bring you to a point where you need serial flashing instead of more convenient OTA updates.

**Reply by nikvostro** [reply]
> How did you get IR and I2C both working at the same time? Currently using Tasmota 10.1.0, and I can get I2C working in standard tasmota.bin, or Fujitsu IR in tasmota_ir.bin, but not both together. Maybe you used tasmota_ircustom, which is now deleted?

I'm did my own build with full IR drivers (as sfromis said in previous message). I have two devices working on 10.0.0.3 and 2021.01.2 custom builds with IR and I2C enabled at the same time.

**Reply by steveboak** [reply]
I'm familiar with compiling and flashing, using a modified Tuya S06Pro with ESP8266. 
My problem is finding the right combination of #defines that enable both I2C (AHT1X) and IR (FUJITSU_AC) at the same time. 

Selecting "tasmota-ir" in platformio_override.ini, and I2C AHT1X device in user_config_override.h (see below), IR receiver correctly decodes Fujitsu remote commands, but I2C seems to be disabled by IR settings.

Compiling for standard tasmota.bin, I2C device works, IR is seen but not decoded as Fujitsu driver is not linked.
```
// Enable I2C for temperature/humidity sensor 
#ifndef USE_I2C
#define USE_I2C
#warning *** Buiding I2C ***
#endif
#define USE_AHT1x                                // [I2cDriver43] Enable AHT10/15 humidity and temperature sensor (I2C address 0x38, 0x39) (+0k8 code)
#define I2CDRIVERS_32_63       0xFFFFFFFF          // Enable I2CDriver32 to I2CDriver63
```

**Reply by sfromis** [reply]
> Selecting "tasmota-ir" in platformio_override.ini

Don't do that. With the standard build process, the "preconfigured" mixes do not allow you to add features. You may [need](https://tasmota.github.io/docs/Firmware-Builds/) `USE_IR_REMOTE_FULL` too

**Reply by steveboak** [reply]
So how to add IR_FULL to a standard build?

**Reply by steveboak** [reply]
Still something missing in my #defines...

Added to user_config_override.h
```
// Enable IR for Fujitsu remote control
#ifndef USE_IR_REMOTE_FULL
#define USE_IR_REMOTE_FULL
#endif
```
Now appears to compile IR code, but final link fails with Error: 
```
undefined reference to `IRsend::sendGC(unsigned short*, unsigned short)'
undefined reference to `IRsend::sendRaw(unsigned short const*, unsigned short, unsigned short)'
etc...
```

**Reply by steveboak** [reply]
If it matters, compiling in an online gitpod environment. Unmodified standard and *-ir variants compile, upload and run with no errors.

**Reply by NicoThien** [reply]
@[nikvostro](https://github.com/nikvostro) Thanks alot. That's exactly what I was looking for.

**Reply by leeza1971** [reply]
> > How did you get IR and I2C both working at the same time? Currently using Tasmota 10.1.0, and I can get I2C working in standard tasmota.bin, or Fujitsu IR in tasmota_ir.bin, but not both together. Maybe you used tasmota_ircustom, which is now deleted?
> 
> I'm did my own build with full IR drivers (as sfromis said in previous message). I have two devices working on 10.0.0.3 and 2021.01.2 custom builds with IR and I2C enabled at the same time.

I did my own build with full IR drivers as you suggested.

How did you make Tasmota WebUI interface showing buttons such as "AC set cool", "AC set heat"?

**Reply by barbudor** [reply]
See command 'WebButton'

**Reply by leeza1971** [reply]
> See command 'WebButton'

Command "Webbutton" is for changing the names shown on the web button.

My question was how to assign function to the Web buttons.

In tasmota, when you assign a relay to a GPIO, a web button appears on UI with ON/OFF label above the button indicating the state of the relay while the button itself has the label "Toggle".

The command "Webbutton" is to change "Toggle" to whatever the user wants.

Here, in the above pic, the web buttons do not have  ON/OFF label and it seems that each button has a certain function like AC set cool", "AC set heat" and so on.

My question would be how to create buttons without "ON/OFF" label above the buttons. I know that assigning a certain function to a button can be achieved by rules...but I could not figure out how to create buttons without "ON/OFF" label thereabove.

**Reply by nikvostro** [reply]
@leeza1971 @barbudor look at >W section

```lua
>D
modenum=-1
btnmodenum=0
mode=""
temp=-1
fan=-1
par=0
irdata=""
acsetheatbtn=0
acsetcoolbtn=0
acsetfanbtn=0
acoffbtn=0
tempinp=24
faninp=0


>B
=>tmpoll 20


>F
if chg[acsetcoolbtn]>0 {
btnmodenum=0
}
if chg[acsetheatbtn]>0 {
btnmodenum=1
}
if chg[acsetfanbtn]>0 {
btnmodenum=2
}
if upd[btnmodenum]>0 {
dp0
print %tempinp% %faninp% %btnmodenum%
+>acset %tempinp%%faninp%%btnmodenum%
}
}
if chg[acoffbtn]>0 {
print acoff
+>acoff
}


>W
nm(18 30 1 tempinp "Temperature (18..30)")
nm(0 5 1 faninp "Fan (0-auto, 5-max)")
bu(acsetcoolbtn "AC cool")bu(acsetheatbtn "AC heat")bu(acsetfanbtn "AC fan")
bu(acoffbtn "AC off")

#ACSET(par)
modenum=par%10
fan=((par%100)-modenum)/10
temp=(par-(fan*10)-modenum)/100
switch modenum
case 0
mode="Cool"
case 1
mode="Heat"
case 2
mode="Fan"
ends  
print %temp% %fan% %modenum% %mode%
->IRHVAC {"Vendor":"FUJITSU_AC","Model":1,"Mode":"%mode%","Power":"On","Temp":%1temp%,"FanSpeed":%0fan%,"SwingV":"Off"}

#ACOFF
print acoff
temp=0
->IRHVAC {"Vendor":"FUJITSU_AC","Model":1,"Power":"Off"}

#ACFANSWING
print AC Fan swing
temp=23
fan=3
Mode="Fan"
->IRHVAC {"Vendor":"FUJITSU_AC","Model":1,"Mode":"Fan","Power":"On","Temp":23,"FanSpeed":3,"SwingV":"Auto"}

#ACCOOL
print AC Cool
=#ACSET(2300)

#ACFREEZE
->IRHVAC {"Vendor":"FUJITSU_AC","Model":1,"Mode":"Cool","Power":"On","Temp":18,"FanSpeed":4,"SwingV":"Auto"}

#ACHOT
print AC Hot
=#ACSET(2531)

#ACHEAT
print AC Hot
=#ACSET(2201)
```

**Reply by BenJamesAndo** [reply]
Newbie here. Where does this code get inserted? Berry Scripting console? Added as a new .json file?

**Reply by sfromis** [reply]
It needs a special build with the scripter feature to support that language. Something similar could be done with Berry.

**Comment by modicon1** [comment]
Do you still have the custom firmware for the ir controller and the AVATTO S06 Pro temp sensor? If so would I be able to download it. I have the identical setup and I modified the board as per your instruction

**Comment by Onepamopa** [comment]
Same here.. Is there a prebuilt version we can use?
Tuya S06 Pro with AHT20 sensor, the chip is BK7231T (WB3S) (no idea if it's supported in Tasmota)

**Reply by sfromis** [reply]
No, that MCU is not supported, but if you *can* swap WB3S for a module which works with Tasmota. This is what was already suggested in the beginning of this thread.

**Reply by modicon1** [reply]
Thanks sfromis. I have already done the hardware mod and it works fine with the tasmota ir firmware loaded but it does not support the I2C temperature sensors. [nikvostro](https://github.com/nikvostro) has modded the firmware to include the temperature/humidity sensor and I was hoping I could just download his modded tasmota firmware.

**Reply by barbudor** [reply]
Moded?

Tasmota is a modular software. That's not modding. Everything is available in the repository. 
Just learn how to build the configuration you need instead of being dependent on others
https://tasmota.github.io/docs/Compile-your-build/

**Reply by modicon1** [reply]
OK barbudor, if you prefer "customized" rather than "modded".

For this single project I was trying to avoid spending a huge amount of time learning how to use Gitpod when the job has already been done by someone else. If that person is willing and able to share, then job done.

**Reply by sfromis** [reply]
Using Gitpod or TasmoCompiler is pretty quick to do. Not much of a learning curve

**Reply by modicon1** [reply]
Can i just start with Tasmota ir and add the sensors i need and re compile?

**Reply by barbudor** [reply]
As written in the doc:
![image](https://github.com/arendst/Tasmota/assets/2996042/313abbe1-e365-439c-b2d1-fe5b4131cd65)

You can't add sensors to tasmota-ir because the definition of the variant explicitly disable I2C

The simplest if to use TasmoCompiler as there is just a few boxes to check to configure a build

If you want to use GitPod or VSCode, then you need to build `tasmota` with the following lines in ùser_config_override.h`
```
#define USE_IR_REMOTE                            // Enable IR remote commands using library IRremoteESP8266
#define USE_IR_REMOTE_FULL                       // Support all IR protocols from IRremoteESP8266
#define USE_AHT1x                              // [I2cDriver43] Enable AHT10/15 humidity and temperature sensor (I2C address 0x38, 0x39) (+0k8 code)

// shrink firmware
#undef USE_LIGHT
#undef USE_ENERGY_SENSOR                         // Disable energy sensors (-14k code)
#undef USE_SPI
#undef USE_DISPLAY                               // Disable support for displays
```

The 3 first line are mandatory to add the support for IR and your T&H sensor
The rest is to help keeping the size of the firmware not too big

**Reply by modicon1** [reply]
Thank you so much barbudor, for sending me that information it helped a great deal. Unfortunately it didn't compile without errors so I'll have to spend sometime debugging. Probably tomorrow before I have time.

**Reply by barbudor** [reply]
It does compile without error on my side
Using latest development branch

**Reply by modicon1** [reply]
I am using tasmota master which loads automatically when I start gitpod. I don’t know how to load different version such as development branch. I have tried a few times using master and it fails with errors after about 30 seconds. 

I have tried tasmocompiler as well. It completes without error but does not work when uploaded.

**Reply by sfromis** [reply]
You can start the Gitpod setup for Tasmota with one URL for the master branch, and a different URL for the development branch.
See https://tasmota.github.io/docs/Gitpod/

**Reply by modicon1** [reply]
This is the bottom part of the error I get;


Linking .pio/build/tasmota/firmware.elf
/home/gitpod/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld: .pio/build/tasmota/src/tasmota.ino.cpp.o:(.text._Z23IrRemoteSendRawStandardPPctjj+0x0): undefined reference to `IRsend::sendRaw(unsigned short const*, unsigned short, unsigned short)'
/home/gitpod/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld: .pio/build/tasmota/src/tasmota.ino.cpp.o: in function `IrRemoteSendRawStandard(char**, unsigned short, unsigned int, unsigned int)':
tasmota.ino.cpp:(.text._Z23IrRemoteSendRawStandardPPctjj+0x91): undefined reference to `IRsend::sendRaw(unsigned short const*, unsigned short, unsigned short)'
/home/gitpod/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld: .pio/build/tasmota/src/tasmota.ino.cpp.o: in function `IrRemoteSendRawFormatted(char**, unsigned int, unsigned int)':
tasmota.ino.cpp:(.text._Z24IrRemoteSendRawFormattedPPcjj+0x131): undefined reference to `IRsend::sendRaw(unsigned short const*, unsigned short, unsigned short)'
/home/gitpod/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld: tasmota.ino.cpp:(.text._Z24IrRemoteSendRawFormattedPPcjj+0x22e): undefined reference to `IRsend::sendRaw(unsigned short const*, unsigned short, unsigned short)'
/home/gitpod/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld: .pio/build/tasmota/src/tasmota.ino.cpp.o:(.text._Z14IrRemoteSendGCPPcjj+0x0): undefined reference to `IRsend::sendGC(unsigned short*, unsigned short)'
/home/gitpod/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld: .pio/build/tasmota/src/tasmota.ino.cpp.o: in function `IrRemoteSendGC(char**, unsigned int, unsigned int)':
tasmota.ino.cpp:(.text._Z14IrRemoteSendGCPPcjj+0x9b): undefined reference to `IRsend::sendGC(unsigned short*, unsigned short)'
collect2: error: ld returned 1 exit status
*** [.pio/build/tasmota/firmware.elf] Error 1
===================================================================================== [FAILED] Took 29.18 seconds =====================================================================================

Environment    Status    Duration
-------------  --------  ------------
tasmota        FAILED    00:00:29.180
================================================================================ 1 failed, 0 succeeded in 00:00:29.180 ================================================================================
gitpod /workspace/Tasmota (master) $ _

**I DO NOT get the error when I remove the changes to the user_config_override.h**

**Reply by modicon1** [reply]
Further to my last post,

If I modify the user_config_override.h with:
**********************************************************************************************************************************
#define USE_IR_REMOTE                            // Enable IR remote commands using library IRremoteESP8266
#define USE_IR_REMOTE_FULL                       // Support all IR protocols from IRremoteESP8266
#define USE_AHT1x                              // [I2cDriver43] Enable AHT10/15 humidity and temperature sensor (I2C address 0x38, 0x39) (+0k8 code)

// shrink firmware
#undef USE_LIGHT
#undef USE_ENERGY_SENSOR                         // Disable energy sensors (-14k code)
#undef USE_SPI
#undef USE_DISPLAY                               // Disable support for displays
*******************************************************************************************
Then I comment out #define USE_IR_REMOTE_FULL                       // Support all IR protocols from IRremoteESP8266
it compiles fine without error. If I leave it in I get an error.

**Reply by barbudor** [reply]
Delete folders `.pio` `.cache` and rebuild
Are you using development branch?

**Reply by modicon1** [reply]
I'm using master branch.

**Reply by sfromis** [reply]
Then switch to the development branch, by using the suitable URL, see the link I posted about that.

**Reply by modicon1** [reply]
Thanks sfromis but I don't see the link

**Reply by sfromis** [reply]
You need new glasses! :eyeglasses: 18 hours ago, I posted:

> You can start the Gitpod setup for Tasmota with one URL for the master branch, and a different URL for the development branch.
See https://tasmota.github.io/docs/Gitpod/

**Reply by modicon1** [reply]
Doh! Thanks man I'm just waking up here and I guess I'm not yet awake!

**Reply by modicon1** [reply]
I ran the compile on the development branch and had the exact same error as I did on master branch.

If I comment out #define USE_IR_REMOTE_FULL it compiles fine without error.

**Reply by modicon1** [reply]
I deleted the .pio and .cache folders as suggested by barbudor and voila... success. Thanks for all your help all of you and with a special shout out to barbudor and sfromis

**Reply by modicon1** [reply]
Hi All,

Can I open this thread again for a related problem?

I complied the new binary and uploaded it to the processor using serial. I then soldered the processor onto the circuit board. Later I made a change and wanted to OTA the change onto the board and the OTA came back with not enough room. I then tried to load Tasmota-minimal and that to came back with "Not enough room". Is there any way with OTA to get around this problem? I don't want to have to unsolder the chip all the time.

**Reply by sfromis** [reply]
You **must** use the compressed `.gz` of the binaries for OTA. Did you do that?

**Reply by modicon1** [reply]
Yes I did for all uploads.

http://ota.tasmota.com/tasmota/release/tasmota-minimal.bin.gz  (261 kb)
tasmota (IR+AHT1x + AHT2x).bin.gz  (522 kb) - This is my custom build.

When doing the custom build, is there any thing that has to be done in the user_config_override to enable OTA?

**Reply by sfromis** [reply]
522 KB compressed is **very** large build, likely over 750 KB. My experience is that a 759 KB build precludes OTA upgrades. Basically the minimal build should fit into what you have in "Free Proram Space" on the Information page in the web UI. Beware of potential confusion of "K" being 1024 or 1000. Current "minimal" build is 261 KB compressed.

There exists a smaller binary usable for OTA, but does IIRC not do an "upgrade", meaning that you may have to configure from scratch again, connecting to the Wifi AP being broadcast.
https://github.com/tasmota/ESP8266MinimalHTTPUpdater

**Reply by modicon1** [reply]
From the INFORMATION page;

Flash Size:  4096 Kb
Program Flash Size:  1024 KB
Program Size:  769 Kb
Free Program Space: 232 Kb
Free Memory 21.5

From what you are saying it sounds like I have no choice but to upgrade using serial.   :-(

Any tips on reducing the size of my custom binary?

**Reply by sfromis** [reply]
Well, the link I posted **is** for such a case, without breaking open and soldering.

**Reply by modicon1** [reply]
Thanks. I'll give it a shot and let you know how I make out.

**Reply by modicon1** [reply]
Well... After I figured out that the updater can't handle .gz files I managed to upload my binary. The problem is that the upload over wrote the WiFi credentials and didn't give an opportunity to reenter them. So now I can't communicate with the device.

Any suggestions before I desolder the chip again?

**Reply by sfromis** [reply]
If you got it done using the right binary, I already mentioned:
> you may have to configure from scratch again, connecting to the Wifi AP being broadcast.

And a `.gz` binary is only for OTA upgrades of Tasmota, not this "emergency procedure". If that was the last one you updated to, no recovery.

**Comment by Onepamopa** [comment]
For some reason I'm unable to flash it, gives an error ..... Everything looks soldered good, but...
Tried using the arduino IDE, getting only:
Serial port COM5
Connecting........_____....._____....._____

**Reply by sfromis** [reply]
Typical reason for that is not having put the chip in flashing mode, grounding gpio0 while applying power.

**Reply by barbudor** [reply]
I also see on the picture above that GPIO15 should be connected to GND
GPIO15 not pulled down at boot is also a common case for not booting the ESP properly

Most dev board have a pull-down on GPIO15 but maybe this board doesn't have.

**Reply by modicon1** [reply]
Is there any way to force an AP mode from this state so I can connect to my network?

**Reply by Onepamopa** [reply]
@sfromis yes.. that's what was causing the issue.

**Reply by sfromis** [reply]
What's possible depends on the meaning of "this state".

**Reply by modicon1** [reply]
I successfully completed the "Updater" and upgraded to my full binary and am left with an updated Tasmota device without wifi credentials.

The binary is a current Development branch with this in the user_config_override.h

#define USE_IR_REMOTE                            // Enable IR remote commands using library IRremoteESP8266
#define USE_IR_REMOTE_FULL                       // Support all IR protocols from IRremoteESP8266
#define USE_AHT1x                              // [I2cDriver43] Enable AHT10/15 humid and temp sensor (I2C address 0x38, 0x39) (+0k8 code)
#define USE_AHT2x                              // [I2cDriver43] Enable AHT20 humid and temp sensor (I2C address 0x38) (+0k8 code)

// shrink firmware
#undef USE_LIGHT
#undef USE_ENERGY_SENSOR                         // Disable energy sensors (-14k code)
#undef USE_SPI
#undef USE_DISPLAY                               // Disable support for displays

**Reply by sfromis** [reply]
You cannot have both AHT1x and AHT2x, the driver is either-or. Should not affect getting it up. As for Wifi credentials, you can bake them in when building, connect to the AP and set them there, or do it via serial commands.

**Reply by modicon1** [reply]
How do I connect to the AP

**Reply by sfromis** [reply]
I'm usually doing it via a phone/tablet.

**Reply by modicon1** [reply]
There is no AP available to connect to so I de-soldered and serial uploaded a new binary with WiFi credentials built in and just the single sensor driver AHT2x.

I re-soldered it into my device and it ran fine except the AHT2x temperature and Humidity labels and values do not show up on the Main Menu page. When I do a I2cScan from the console, it shows {"I2CScan":"Device(s) found at 0x38"} but there are no SENSOR MQTT statements.

Should I have added anything else to the user_config_override when I built the binary?
Any other ideas?

Thanks

**Reply by barbudor** [reply]
So you're back to a tasmota configuration question
What is the result of command I2CDRIVER ?
It should
- List the I2C drivers that are built-in the binary, and show those that are enabled. AHT driver is 43. => If 43 is not there, your biuld is not correct
- There should be no exclamation mark in front of 43 => If there is !43, enter command `I2CDriver43 1`

Preferably every driver number that you don't need should be disabled

**Reply by modicon1** [reply]
Thanks for responding barbudor. The command I2CDRIVER returns {"I2CDriver":"43"}

**Reply by sfromis** [reply]
Are you positive that is it is actually an AHT20 or AHT25? If is an AHT10 (as in the device in the start of this thread), the sensor will AFAICT not respond if you use a build including `USE_AHT2x` as you cannot have both types included at the same time.

**Reply by barbudor** [reply]
Can you try the following change for more logs on the 2 functions below ?
(tasmota\tasmota_xsns_sensor\xsns_63_aht1x.ino)
That should help to understand where it fails
```C++
bool AHT1XInit(uint8_t aht1x_address) {
  AddLog(0,PSTR("AHT: try 0x%02X"), aht1x_address);
  Wire.beginTransmission(aht1x_address);
  Wire.write(AHTSetCalCmd, 3);
  uint8_t res = Wire.endTransmission();
  AddLog(0,PSTR("AHT: res1=0x%02X"), res);
  if (res != 0)
    return false;
  delay(AHT1X_CMD_DELAY);
  res = AHT1XReadStatus(aht1x_address);
  AddLog(0,PSTR("AHT: res2=0x%02X"), res);
  if(res & 0x08) // Sensor calibrated?
    return true;
  return false;
}

void AHT1XDetect(void) {
  for (uint8_t i = 0; i < AHT1X_MAX_SENSORS; i++) {
    if (!I2cSetDevice(aht1x.addresses[i])) {continue;}
    if (AHT1XInit(aht1x.addresses[i])) {
      AddLog(0,PSTR("AHT: %s at 0x%02X"), aht1x_sensors[aht1x.count].types, aht1x_sensors[aht1x.count].address);
      aht1x_sensors[aht1x.count].address = aht1x.addresses[i];
      GetTextIndexed(aht1x_sensors[aht1x.count].types, sizeof(aht1x_sensors[aht1x.count].types), i, ahtTypes);
      I2cSetActiveFound(aht1x_sensors[aht1x.count].address, aht1x_sensors[aht1x.count].types);
      aht1x.count++;
      break; // Only one Sensor allowed at the moment (I2C Sensor-Bug)
    }
  }
}
```

**Reply by modicon1** [reply]
Thanks for responding sfromis. Yes I'm sure. The device I'm using is a S06Pro IR Controller and the Temperature sensor is clearly marked AHT20. The binary is compiled with #define USE_AHT2x.

I have another IR Controller that is also a S06Pro but it has an AHT10 sensor and it works fine (different custom binary compiled with #define USE_AHT1x)

**Reply by barbudor** [reply]
I would assume that you tested the AHT1x binary that works on the other controller on that controller ? 
Just in case the marking on the chip is random..

**Reply by barbudor** [reply]
Thus, while lookink at the code, the only dependency of the driver on USE_AHT1x/USE_AHT2x is the string that display the sensor type...

**Reply by modicon1** [reply]
Sorry barbudor you lost me on the change you wanted me to make. Where does that code go? It is deeper than I have gone before.

**Reply by barbudor** [reply]
As written : tasmota\tasmota_xsns_sensor\xsns_63_aht1x.ino
Find those functions in that module (should start at line 150)
Delete them
Use the code I have pasted

No change in the code, but some more logs that you should see at restart

**Reply by modicon1** [reply]
What are we using to view the code.

**Reply by sfromis** [reply]
> Thus, while lookink at the code, the only dependency of the driver on USE_AHT1x/USE_AHT2x is the string that display the sensor type...

The more important [difference](https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xsns_sensor/xsns_63_aht1x.ino#L63-L72) is the value for `#define AHTX_CMD`

**Reply by modicon1** [reply]
OK barbudor I got you. You want me recompile the firmware with the new code.

**Reply by modicon1** [reply]
I'm sorry but I don't think I'm as smart as you think I am. :-)

**Reply by barbudor** [reply]
But as you have been able to compile a binary wiht AHT, I assumed your were ok for such change
if that's over you skills, let me know and I'll provide you with a test binary

**Reply by modicon1** [reply]
barbudor thanks, but I think I'm OK building the binary. I have learned a lot from you and sfromis in the last few days. My big concern is updating the device OTA. You remember what we went through uploading the original large binary. :-(

I'll let you know how it goes. I won't be able to do anything else for a while (probably tomorrow) but I'll get back to you.

**Reply by barbudor** [reply]
I've build one, small enough to be flashed in 1 single step using file upload
[tasmota-test.zip](https://github.com/arendst/Tasmota/files/12910137/tasmota-test.zip)

**Reply by modicon1** [reply]
Hi barbudor, I just tried uploading your test file and it failed "Not enough space"

**Reply by modicon1** [reply]
I just noticed something while I was looking around this morning... Even though the AHT2x has a driver listed and able to activate, the AHT20 is not shown in the Tasmota list of supported sensors. Do you think that has something to do with my problem?

**Reply by sfromis** [reply]
I'm using AHT25 just fine with `USE_AHT2x` configured.
`"AHT2X":{"Temperature":21.79,"Humidity":47.69,"DewPoint":10.17}`

Not sure what you mean by "the Tasmota list of supported sensors". I'm seeing it in all of:
https://tasmota.github.io/docs/I2CDEVICES/#supported-i2c-devices
https://tasmota.github.io/docs/Supported-Peripherals/
https://tasmota.github.io/docs/AHT2x/

If what you think of is the dropdown when configuring a GPIO pin, I2C devices will not appear there, as the sensor type is not a pin configuration when it is on the I2C bus.

To upgrade to a new binary via file upload, you need to first upgrade to `tasmota-minimal.bin.gz` as a stepping stone.

**Reply by modicon1** [reply]
OK, so not my problem. I didn't see all those. I really do need new glasses. :-)

As for the upload... barbudor suggested I could do it in 1 step but I will try minimal first.

**Reply by modicon1** [reply]
sfromis, Minimal will not upload either... "Not enough space"

**Reply by barbudor** [reply]
Are you using the GZ file ?

**Reply by modicon1** [reply]
Yes

**Reply by sfromis** [reply]
In that case, again https://github.com/tasmota/ESP8266MinimalHTTPUpdater
Make sure to not use the `.gz` binary for **that**. And not minimal, either.

**Reply by modicon1** [reply]
Updater doesn't do the job for me. As soon as I upload the firmware I need from updater I lose the wifi credentials and can't log back in to the device unless it has an AP which, as far as I can determine, it doesn't.

**Reply by sfromis** [reply]
Yes, it **will** lose wifi credentials, and Tasmota should provide an AP, assuming that you supply enough power for that to work.

**Reply by modicon1** [reply]
How do I call AP mode from Tasmota?

**Reply by sfromis** [reply]
In this context, you don't. More generally - depends on what you want, those short words does not tell that.

I just tested the tool I suggested. Downloaded both `updater.bin` and `tasmota.bin` to a mobile device, used Tasmota upgrade to load `updater.bin`, and then connected to the AP "UpdateMe", password "recovery", where I then could select my `tasmota.bin` and perform the update. Worked without a hitch. Did not even lose my Config.

**Reply by modicon1** [reply]
When you updated to tasmota.bin did you retain your wifi settings? I used Updater as you described and it work for me as well. The problem was when I updated the Tasmota firmware to tasmota.bin from updater I lost my wifi credentials and could no longer communicate with the device using wifi.

**Reply by sfromis** [reply]
Yes, "did not even lose my config" includes Wifi. But if the config is lost, Tasmota is usually able to go into initial config mode, broadcasting an AP to connect to.

**Reply by modicon1** [reply]
So Tasmota automatically broadcasts an AP SSID when it loses wifi config?

**Reply by sfromis** [reply]
If it loses config as such, meaning not finding a settings data structure with a correct checksum, it should treat it as a fresh flash, and do a full initialization, which includes the standard AP mode. The binary tasmota-minimal can't do this, but needs a working config already.

If it for some reason has bad wifi setup, the usual [recovery](https://tasmota.github.io/docs/Device-Recovery/) techniques might be relevant.

**Reply by Onepamopa** [reply]
Do note, on these S06 Pro's ... the Tuya mcu needs to be removed before initially flashing the ESP, else it won't flash. You can re-solder it back after initial flash, it's useable.

**Comment by abuhabeeb** [comment]
I have S06 Pro with AHT20 sensor, I mange to work it with ESPHome it shows the  temperature and Humidity, I tried with Tasmota but labels and values do not show up .
When I do a I2cScan from the console, it shows {"I2CScan":"Device(s) found at 0x38"} 
The command I2CDRIVER returns {"I2CDriver":"43"}

Added to user_config_override.h

```
#ifndef USE_AHT2x
#define USE_AHT2x       Enable AHT20 instead of AHT1x humidity and temperature sensor (I2C address 0x38)
#endif

#define USE_IR_REMOTE                            // Enable IR remote commands using library IRremoteESP8266
#define USE_IR_REMOTE_FULL                       // Support all IR protocols from IRremoteESP8266
``` 

Anything else I miss?

@modicon1 Did you manage to make it work with your S06 Pro with AHT20 sensor?

**Reply by modicon1** [reply]
Abuhabeeb I did everything as you described except for esphome but never got it to work with the AHT20. After much back and forth with the team I finally gave up. When I get back home in about a week, I am going to swap the AHT20 for an AHT10 and be done with it.

**Reply by modicon1** [reply]
I suspect there might be a bug with the AHT20 sensor driver. I physically replaced the AHT20 with an AHT10 sensor, updated the .bin to #define USE_AHT1x instead of #define USE_AHT2x and Humidity, Temperature, and Dewpoint showed up on the Tasmota UI as expected.

**Reply by excodex** [reply]
@abuhabeeb hey, can you please share your esphome config? thanks!

**Comment by marsman7** [comment]
I have also converted the Avatto PCB to ESP8266 and it also works. By default, the operating voltage of the AHT10 sensor is switched via transistor Q5. In the description above, this is bypassed. Does Tasmota offer a way to switch the sensor on, retrieve data and switch it off again? This minimizes the self-heating of the sensor.

**Reply by sfromis** [reply]
No, but the "problem" with heat and sensors is usually not the sensor, but the Wifi device generating heat, spreading inside the case to skew what the sensor gets.
